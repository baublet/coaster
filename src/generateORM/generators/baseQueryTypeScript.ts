import { ConnectionOptions } from "../";
import { MetaData } from ".";
import { generateNames } from "../../generateNames";
import { RawSchema } from "../drivers";
import { getSchemaAndTablePath } from "./helpers";

/**
 * Generates the lowest-level possible accessor for accessing data in a table
 * using raw database types.
 */
export const baseQueryTypeScript = (
  schema: RawSchema,
  metaData: MetaData,
  options: {
    knexConnectionOptions?: ConnectionOptions;
  } = { knexConnectionOptions: {} }
) => {
  const connectionOptions = options.knexConnectionOptions
    ? JSON.stringify(options.knexConnectionOptions)
    : "";
  let code = `import knex from "knex";\n\n`;
  code += `export const getDatabaseConnection = () => knex(${connectionOptions});\n\n`;

  for (const table of schema.tables) {
    const entityName = metaData.tableEntityNames.get(
      getSchemaAndTablePath(schema.name, table.name)
    );
    const pluralEntityName = generateNames(entityName).pluralPascal;
    code += `export function ${pluralEntityName}<Result = ${entityName}[]>(
  connection = getDatabaseConnection()
) {
  return connection<${entityName}, Result>("${table.name}");
};\n\n`;
  }

  return code;
};
