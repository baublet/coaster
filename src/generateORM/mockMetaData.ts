import { MetaData } from "./generators";
import { getTemplateManager } from "../generateORM/generators/helpers";

export function getMockMetaData(): MetaData {
  return {
    testConnectionVariable: "connection",
    codeOutputFullPath: "./generated.ts",
    generateTestCode: false,
    entityTableNames: new Map(),
    setHeader: jest.fn(),
    setTestHeader: jest.fn(),
    tableEntityNames: new Map(),
    rawBaseQueryFunctionNames: new Map(),
    transformerFunctionNames: {},
    typeAssertionFunctionNames: new Map(),
    typeGuardFunctionNames: new Map(),
    tableRawEntityNames: new Map(),
    namedEntityColumnNames: new Map(),
    namedEntityInputTypeNames: new Map(),
    namedDatabaseEnumNames: new Map(),
    rawDatabaseEnumNames: new Map(),
    namedCreateTestEntityFunctionNames: new Map(),
    rawCreateTestEntityFunctionNames: new Map(),
    rawEnumValues: new Map(),
    templateManager: getTemplateManager(),
    namedMockEntityFunctionNames: new Map(),
    rawMockEntityFunctionNames: new Map(),
  };
}
