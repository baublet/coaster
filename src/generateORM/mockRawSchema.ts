import { RawSchema } from "./drivers";

export const getMockRawSchema: () => RawSchema = () => ({
  name: "public",
  flavor: "pg",
  enums: [
    {
      name: "user_account_source",
      values: ["github", "bitbucket"],
    },
  ],
  tables: [
    {
      name: "user",
      primaryKeyColumn: "id",
      primaryKeyType: "string",
      uniqueConstraints: [],
      columns: [
        {
          name: "id",
          columnType: "columnType",
          nullable: false,
          type: "string",
          foreignKeys: [],
          hasDefault: true,
        },
        {
          name: "name",
          comment: "Users don't have to provide their name",
          columnType: "columnType",
          nullable: true,
          type: "string",
          foreignKeys: [],
          hasDefault: false,
        },
        {
          name: "data",
          columnType: "columnType",
          nullable: false,
          type: "AnyJson",
          foreignKeys: [],
          hasDefault: false,
        },
      ],
    },
    {
      name: "user_account",
      comment: "Singular users can have multiple accounts",
      primaryKeyColumn: "id",
      primaryKeyType: "string",
      uniqueConstraints: [],
      columns: [
        {
          name: "id",
          columnType: "columnType",
          nullable: false,
          type: "unknown",
          foreignKeys: [],
          hasDefault: true,
        },
        {
          name: "source",
          comment: "Who sent us this users' data?",
          columnType: "columnType",
          nullable: false,
          type: "enum",
          foreignKeys: [],
          hasDefault: false,
          enumPath: "public.user_account_source",
        },
        {
          name: "source_data",
          comment: "The raw data sent from the source provider",
          columnType: "columnType",
          nullable: false,
          type: "AnyJson",
          foreignKeys: [],
          hasDefault: false,
        },
      ],
    },
  ],
});
