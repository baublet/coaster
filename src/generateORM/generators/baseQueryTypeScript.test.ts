import { baseQueryTypeScript } from "./baseQueryTypeScript";
import { mockRawSchema } from "../mockRawSchema";
import { getMockMetaData } from "../mockMetaData";

const metaData = getMockMetaData();

it("generates a schema", () => {
  metaData.tableEntityNames.set("public.user", "RawUser");
  metaData.tableEntityNames.set("public.userAccount", "RawUserAccount");
  expect(baseQueryTypeScript(mockRawSchema, metaData)).toMatchSnapshot();
});
