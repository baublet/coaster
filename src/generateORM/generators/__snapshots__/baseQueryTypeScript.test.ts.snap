// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generates a schema 1`] = `
"export const getPublicConnection = () => knex();

export function RawUsers<Result = RawUser[]>(
  options: options = { connection: getPublicConnection() }
) {
  return options.connection<RawUser, Result>(\\"user\\");
};

export function RawUserAccounts<Result = RawUserAccount[]>(
  options: options = { connection: getPublicConnection() }
) {
  return options.connection<RawUserAccount, Result>(\\"user_account\\");
};

"
`;
